purchase_update_proposal/models/purchase_order.py

```diff
@@ -0,0 +1,294 @@
+                        else:
+                            prevent_update = prevent_update or False
+                if prevent_update:
+                    rec.proposal_updatable = "no"
+                else:
+                    rec.proposal_updatable = "yes"
+            else:
+                rec.proposal_updatable = "yes"
+
+    def _compute_proposal(self):
+        for rec in self:
+            rec.proposal_count = len(rec.proposal_ids)
+
+    def populate_all_purchase_lines(self):
+        self.ensure_one()
+        lines = self.order_line.filtered(
+            lambda s: s.id not in self.proposal_ids.mapped("line_id").ids
+            and s.received != "all"
+        )
+        self.write({"proposal_date": False})
+        lines.button_update_proposal()
+
+    def submit_proposal(self):
+        self.ensure_one()
+        self._prepare_proposal_data()
+        self._check_updatable_proposal()
+        if self.proposal_updatable == "yes":
+            self.write({"proposal_state": "submitted"})
+
+    def reset_proposal(self):
+        self.ensure_one()
+        self.write({"proposal_state": "draft"})
+        self._check_updatable_proposal()
+
+    def reject_proposal(self):
+        self.ensure_one()
+        self.write({"proposal_state": "rejected"})
+        self._check_updatable_proposal()
+
+    def approve_proposal(self):
+        """
+        We need to update purchase lines with approved lines.
+        If 1 purchase line has created more than 1 line,
+        we update the first one and create another purchase line
+        """
+        self.ensure_one()
+        if not self._get_purchase_groups():
+            raise UserError(_("You are not authorized to approve this proposal"))
+        self._check_updatable_proposal()
+        if self.proposal_updatable == "no":
+            # example: qty is in proposal and purchase is partially received
+            return
+        body = []
+        initial_state = False
+        # these proposals'll reset these lines
+        lines_to_0 = self.proposal_ids.filtered(lambda s: s.qty == 0.0).mapped(
+            "line_id"
+        )
+        data = self._prepare_proposal_data()
+        if data:
+            self._hook_pending_proposal_approval()
+            if self._product_qty_key_in_data(data) and self.state in [
+                "confirmed",
+                "approved",
+            ]:
+                # We have to reset order because of changed qty in confirmed order
+                initial_state = self.state
+                self.action_cancel()
+                self.action_cancel_draft()
+            self._update_proposal_to_purchase_line(data, body)
+            self.message_post(body="\n".join(body))
+        if initial_state in ("approved", "confirmed"):
+            # altered quantity implied to reset order, then we approve them again
+            self.signal_workflow("purchase_confirm")
+            self.signal_workflow("purchase_approve")
```

Comment:
These 2 lines above shouldn't exists because it's 8.0 syntax.

Also in 16.0 we shouldn't cancel an order to update it. It should impact many lines of code